File: platform/src/main/java/ru/abondin/hreasy/platform/BusinessError.java
package ru.abondin.hreasy.platform;


import lombok.Data;

import java.util.Map;

/**
 * Business Error Exception.
 * Give the client clear reason of error.
 */
@Data
public class BusinessError extends RuntimeException {


    /**
     * Try to use methods from {@link BusinessErrorFactory} instead of directly using this constructor
     * @param code
     * @param localizationArgs
     */
    public BusinessError(String code, String... localizationArgs) {
        this.code = code;
        this.localizationArgs = localizationArgs;
    }

    /**
     * Error code.
     * Uses in localization process.
     * Sends to the client side
     */
    private String code;
    /**
     * Arguments to resolve full message from the localization template
     */
    String[] localizationArgs;
    /**
     * Additional attributes to send to the client
     */
    Map<String, String> attrs;
    /**
     * Default message if no localization found
     */
    String defaultMessage;


}

File: platform/src/main/kotlin/ru/abondin/hreasy/platform/BusinessError.kt
package ru.abondin.hreasy.platform

/**
 * Business Error Exception.
 * Give the client clear reason of error.
 */
class BusinessError(
        /**
         * Error code.
         * Uses in localization process.
         * Sends to the client side
         */
        val code: String,
        /**
         * Arguments to resolve full message from the localization template
         */
        val localizationArgs: Array<out Any>,
        /**
         * Additional attributes to send to the client
         */
        val attrs: Map<String, String>,
        /**
         * Default message if no localization found
         */
        val defaultMessage: String?,
        /**
         * Business Error Cause
         */
        cause: Exception?) : RuntimeException(defaultMessage, cause) {
    constructor(code: String, localizationArgs: Array<out Any>, cause: Exception?) : this(code, localizationArgs, mapOf(), null, cause);
    constructor(code: String, cause: Exception?) : this(code, arrayOf(), cause);
    constructor(code: String, localizationArgs: Array<out Any>) : this(code, localizationArgs, null)
 }